ui-v2/app/services/blocking.js

```diff
@@ -0,0 +1,91 @@
+const usage = new MultiMap(Set);
+
+export default Service.extend({
+  client: service('client/http'),
+  settings: service('settings'),
+
+  finder: service('repository/manager'),
+
+  open: function(uri, ref) {
+    let source;
+    // Check the cache for an EventSource that is already being used
+    // for this uri. If we don't have one, set one up.
+    if (!sources.has(uri)) {
+      // Setting up involves finding the correct repository method to call
+      // based on the uri, we use the eventually injectable finder for this.
+      const repo = this.finder.fromURI(...uri.split('?filter='));
+      // We then check the to see if this we have previously cached information
+      // for the URI. This data comes from caching this data when the EventSource
+      // is closed and destroyed. We recreate the EventSource from the data from the cache
+      // if so. The data is currently the cursor position and the last received data.
+      let configuration = {};
+      if (cache.has(uri)) {
+        configuration = cache.get(uri);
+      }
+      // tag on the custom createEvent for ember-data
+      configuration.createEvent = repo.reconcile;
+
+      // We create the EventSource, checking to make sure whether we should close the
+      // EventSource on first response (depending on the user setting) and reopening
+      // the EventSource if it has been paused by the user navigating to another tab
+      source = new EventSource((configuration, source) => {
+        const close = source.close.bind(source);
+        const deleteCursor = () => (configuration.cursor = undefined);
+        //
+        return maybeCall(deleteCursor, ifNotBlocking(this.settings))().then(() => {
+          return repo
+            .find(configuration)
+            .then(maybeCall(close, ifNotBlocking(this.settings)))
+            .catch(restartWhenAvailable(this.client));
+        });
+      }, configuration);
+      // Lastly, when the EventSource is no longer needed, cache its information
+      // for when we next need it again (see above re: data cache)
+      source.addEventListener('close', function close(e) {
+        const source = e.target;
+        source.removeEventListener('close', close);
+        cache.set(uri, {
+          currentEvent: e.target.getCurrentEvent(),
+          cursor: e.target.configuration.cursor,
+        });
+        // the data is cached delete the EventSource
+        sources.delete(uri);
+      });
+      sources.set(uri, source);
+    } else {
+      source = sources.get(uri);
+    }
+    // set/increase the usage counter
+    usage.set(source, ref);
+    source.open();
+    return source;
+  },
+  close: function(source, ref) {
+    if (source) {
+      // decrease the usage counter
+      usage.remove(source, ref);
+      // if the EventSource is no longer being used
+      // close it (data caching is dealt with by the above 'close' event listener)
+      if (!usage.has(source)) {
```

Comment:
> Would you ever have multiple sources with the same key?

The sources themselves are the key, so if you have multiple sources they are all used as a unique key.

> I don't see the size/count exposed anywhere, what's the plan for that?

So to go into this slightly, as you said this is all about counting usage. We count the usage using a `MultiMap`, The keys of the MultiMap are the sources, and the values are the things that are using the source (in this case the values are ember components). So we essentially maintain a record of the amount of things that are using the source. When the MultiMap no longer has a certain source/key (`.has(source)`) we know it is no longer being used by any component/value, in which case we can stop/close/destroy it. 

We use a `Set` (rather than an array) for the container for the values of the `MultiMap` (see further up above) so theoretically you cannot have a component 'registered' more than once with the same source. You've actually got me questioning here why I didn't use a `MultiSet` instead 🤔 so will have a re-look at this....Actually I suppose we don't _count_ as such, we just check if a key exists or not i.e. is count zero or not? Maybe thats why I chose a `MultiMap(Set)` and not a `MultiSet`, my head's not in this right now, but I think that might have been it.

A longer story/plan related to other PRs is, I previously had this as some hand-rolled counting code and chose to bring in `mnemonist` instead for this as I figured I might be using other `mnemonist` and related things (graphing) for later work. That's has not happened as yet (and I've chosen a different graph lib for the disco-chain work). But I think `mnemonist` is full of useful goodies and its nicely packaged so I think I'll probably keep it in here.